datafusion/core/src/physical_optimizer/join_selection.rs

```diff
@@ -157,7 +157,9 @@ fn swap_join_projection(
 }
 
 /// This function swaps the inputs of the given join operator.
-fn swap_hash_join(
+/// This function is public so other downstream projects can use it
+/// to construct `HashJoinExec` with right side as the build side.
+pub fn swap_hash_join(
```

Comment:
Spark supports build right in HashJoin operator internally. In planning stage, it decides which side is build side.

However, DataFusion HashJoin planning takes alternative approach. DataFusion HashJoin operator only supports build left. In optimization stage, once DataFusion finds build right is better, it will call this function to swap inputs for HashJoin to achieve build right effect.

So, to make Comet to use DataFusion HashJoin to support Spark build right HashJoin, we only need this function to be public so Comet can directly use it.

Reply:
Given this input takes `&HashJoinExec` as input, it seems like it might be easier to find / make more sense as a method on `HashJoinExec` itself 🤔

Reply:
+1

Reply:
Sounds good. 👍 
Let me update it accordingly.

Reply:
Hmm, some functions used in `swap_hash_join` are also used by other functions like `swap_nl_join`. So I cannot simply move these functions into `HashJoinExec` as private ones.

Only I can do is move these functions to joins/utils.rs in `physical-plan` crate and make them as public APIs so both join_selection.rs and `HashJoinExec` can use them.

It makes more APIs public actually, including `swap_join_filter`, `swap_reverting_projection` and `swap_join_type`.

I think it is not so worth making more public APIs just for `swap_hash_join`.

Seems the current approach is better?

WDYT?